internal/jujucommands/bootstrap_test.go

```diff
@@ -0,0 +1,256 @@
+	jujucloud "github.com/juju/juju/cloud"
+
+	"github.com/canonical/jimm/v3/internal/jujucommands"
+)
+
+type runnerMock struct {
+	impl    func(ctx context.Context, args []string) (<-chan jujucommands.OutputLine, error)
+	dataDir string
+}
+
+func (r *runnerMock) RunJujuCmd(ctx context.Context, args []string) (<-chan jujucommands.OutputLine, error) {
+	return r.impl(ctx, args)
+}
+
+func (r *runnerMock) JujuDataDir() string {
+	return r.dataDir
+}
+
+func (s *jujucommandsSuite) TestBootstrapCmdParams_Validate(c *qt.C) {
+	p := jujucommands.BootstrapCmdParams{}
+
+	c.Assert(p.Validate(), qt.ErrorMatches, ".*cloud \\[and region\\] name cannot be empty.*")
+
+	p.CloudNameAndRegion = "testregion/testcloud"
+
+	c.Assert(p.Validate(), qt.ErrorMatches, ".*controller name name cannot be empty.*")
+
+	p.ControllerName = "my-controller"
+
+	c.Assert(p.Validate(), qt.ErrorMatches, ".*login-token-refresh-url cannot be empty.*")
+
+	p.LoginTokenRefreshURL = "myurl.com"
+
+	c.Assert(p.Validate(), qt.IsNil)
+
+	p.AgentVersion = "bad version"
+
+	c.Assert(p.Validate(), qt.ErrorMatches, "invalid version \"bad version\"")
+
+	p.AgentVersion = "1.1.1"
+
+	c.Assert(p.Validate(), qt.IsNil)
+
+	p.BootstrapTimeout = -1
+
+	c.Assert(p.Validate(), qt.ErrorMatches, "bootstrap timeout cannot be less than or equal to 0")
+
+	p.BootstrapTimeout = 1
+
+	c.Assert(p.Validate(), qt.IsNil)
+}
+
+func (s *jujucommandsSuite) TestBootstrapCmdParams_BuildBootstrapCmdArgs(c *qt.C) {
+	p := jujucommands.BootstrapCmdParams{
+		CloudNameAndRegion:   "testregion/testcloud",
+		ControllerName:       "my-controller",
+		AgentVersion:         "1.1.1",
+		BootstrapTimeout:     1000,
+		LoginTokenRefreshURL: "myurl.com",
+	}
+
+	args := p.BuildBootstrapCmdArgs()
+
+	c.Assert(
+		args,
+		qt.DeepEquals,
+		[]string{
+			"bootstrap",
+			"--config",
+			"login-token-refresh-url=myurl.com",
+			"--agent-version=1.1.1",
+			"--config",
+			"bootstrap-timeout=1000",
+			"testregion/testcloud",
+			"my-controller",
+		},
+	)
+
+	p.AgentVersion = ""
+	args = p.BuildBootstrapCmdArgs()
+
+	c.Assert(
+		args,
+		qt.DeepEquals,
+		[]string{
+			"bootstrap",
+			"--config",
+			"login-token-refresh-url=myurl.com",
+			"--config",
+			"bootstrap-timeout=1000",
+			"testregion/testcloud",
+			"my-controller",
+		},
+	)
+
+	p.BootstrapTimeout = 0
+	args = p.BuildBootstrapCmdArgs()
+
+	c.Assert(
+		args,
+		qt.DeepEquals,
+		[]string{
+			"bootstrap",
+			"--config",
+			"login-token-refresh-url=myurl.com",
+			"testregion/testcloud",
+			"my-controller",
+		},
+	)
+}
+
+func (s *jujucommandsSuite) TestBootstrapCmdParams_RunBootstrapCmd_PersonalCloudWritten(c *qt.C) {
+	testCtx := c.Context()
+
+	mock := runnerMock{
+		impl: func(ctx context.Context, args []string) (<-chan jujucommands.OutputLine, error) {
+			// Return chan that has one line inside
+			outputCh := make(chan jujucommands.OutputLine, 1)
+			outputCh <- jujucommands.OutputLine{
+				Line: "testing",
+			}
+			close(outputCh)
```

Comment:
You could simplify this to,

```suggestion
			outputCh := make(chan jujucommands.OutputLine)
			close(outputCh)
```
The same for the other tests.